### PR TITLE
Cluster code a number of fixes and enhancements

### PR DESCRIPTION
**Pull request by naiming-zededa**
- fix a type in applying longhorn disk config label
- k3s yaml file may change after joining the cluster, need to monitor that
- fix a number of issue in transition to cluster in cluster-init.sh
- redo the save /var/lib files and bring them back in single-node mode
- fix the problem of checking bootstrap server status query
- redo the saving /var/lib by terminate k3s first
- add check cluster-config change in the loop even our node is not up, when the 2nd node is removed, the 3rd node will be down, and it still need to check the cluster configure status
- fix a bug in parsing the controller EdgeNodeClusterConfig during removing
- make the EdgeNodeClusterConfig to be persistent
- before we convert from cluster into single-node, we have to first drain and delete our own node before the cluster interface gets shut
- move the zedkube subEdgeNodeClusterConfig Activate to post the acquired DeviceNetworkStatus, etc
- remove the clusterCIDR code in the linux.go, it does not seemt o be needed
- re-eanable the normal inter-device forwarding blocking code, and added a number of cluster related addresses, ports, to allow them
- updated eve-api to get the newly defined Cluster token cipher block defs

### CODE REVIEW

File: pkg/kube/cluster-init.sh [hunk 1]
```diff
@@ -734,6 +715,35 @@ check_cluster_config_change() {
...
+  # Remove everything in the destination directory
+  rm -rf "${dest_dir:?}"/*
+
+  # Copy all contents from /var/lib to destination directory
+  cp -a /var/lib/. "$dest_dir"
+}
+
+# Function to restore contents from /persist/kube/log/save-var-lib back to /var/lib
+restore_var_lib() {
+  local source_dir="/persist/kube/log/save-var-lib"
```
Comment (andrewd-zededa) [inline pkg/kube/cluster-init.sh:735]: @naiming-zededa this source_dir doesn't match dest_dir above in save_var_lib()

File: pkg/kube/cluster-init.sh [hunk 2]
```diff
@@ -283,18 +283,30 @@ apply_node_uuid_lable () {
...
                 touch /var/lib/node-labels-initialized
         else
-                logmsg "Failed to re-apply node labels"
+                logmsg "Failed to re-apply node labels, on $HOSTNAME, uuid $DEVUUID"
         fi
 }
 
+# when transitioning from single node to cluster mode, the k3s.yaml file may need
+# to change with new certificates
+check_kubeconfig_yaml_files() {
```
Comment (zedi-pramodh) [inline pkg/kube/cluster-init.sh:300]: Do we really need to cmp and copy. Why can't we blindly copy whenever we switch to cluster mode ?

File: pkg/kube/cluster-init.sh [hunk 3]
```diff
@@ -721,6 +690,18 @@ check_cluster_config_change() {
...
+
+            if [ "$is_bootstrap" = "true" ]; then
+              # provision this node to cluster mode with bootastrap config
+              logmsg "privision config file for bootstrap node"
+              provision_cluster_config_file true
+            else
+              # provision this node to cluster mode with server config, will wait for join-ip to be ready
+              # we transition to cluster mode, and we need to wait for the bootstrap server status
+              logmsg "privision config file for non-bootstrap node"
+              provision_cluster_config_file true
```
Comment (andrewd-zededa) [inline pkg/kube/cluster-init.sh:702]: @naiming-zededa Why is provision_cluster_config_file in the condition if its called with the same arguments either way?

File: pkg/kube/cluster-init.sh (hunk 2 above)
Comment (naiming-zededa) [reply pkg/kube/cluster-init.sh:300]: No this is not during the swtichover time, when switch from single node to cluster, it actually generated the k3s.yaml twice, one is initially which we does the copy as normal, then after joined the cluster, it gets it from the bootstrap server. So, we don't know when this is happening (well, within some 30 seconds).

File: pkg/kube/cluster-init.sh (hunk 3 above)
Comment (naiming-zededa) [reply pkg/kube/cluster-init.sh:702]: good catch, removed the condition. It was taken different params before, but that was not correct.

File: pkg/kube/cluster-init.sh (hunk 1 above)
Comment (naiming-zededa) [reply pkg/kube/cluster-init.sh:735]: will correct that.

File: pkg/pillar/cmd/zedkube/kubeinterface.go [hunk 4]
```diff
@@ -78,6 +83,116 @@ func runKubeConfig(ctx *zedkubeContext, config, oldconfig *types.EdgeNodeCluster
...
+			return err
+		}
+	}
+
+	return nil
+}
+
+func isDaemonSet(pod *corev1.Pod) bool {
+	// This is a simplified check. A real implementation would need to check the owner references.
+	return pod.Spec.HostNetwork
```
Comment (andrewd-zededa) [inline pkg/pillar/cmd/zedkube/kubeinterface.go:129]: Just a suggestion:
```suggestion
	if pod.OwnerReferences != nil && len(pod.OwnerReferences) > 0 {
		for _, ref := range pod.OwnerReferences {
			if ref.Kind == "DaemonSet" {
			     return true  
			}
		}
	}
	return false
```

File: pkg/pillar/cmd/zedkube/kubeinterface.go [hunk 5]
```diff
@@ -78,6 +83,116 @@ func runKubeConfig(ctx *zedkubeContext, config, oldconfig *types.EdgeNodeCluster
...
+	GracePeriodSeconds  int64
+	Timeout             time.Duration
+	Out                 *os.File
+	ErrOut              *os.File
+}
+
+func RunNodeDrain(drainer *Drainer, nodeName string) error {
+	pods, err := drainer.Client.CoreV1().Pods("").List(context.Background(), metav1.ListOptions{
+		FieldSelector: "spec.nodeName=" + nodeName,
+	})
```
Comment (andrewd-zededa) [inline pkg/pillar/cmd/zedkube/kubeinterface.go:100]: Can we open this up to allow a pod selector string allowing the kubernetes style label syntax (eg. "app!=some-pod-app-name")?  The upcoming upgrade path will make use of something like this.
```suggestion
type Drainer struct {
	Client              *kubernetes.Clientset
	Force               bool
	IgnoreAllDaemonSets bool
	DeleteEmptyDirData  bool
	GracePeriodSeconds  int64
	Timeout             time.Duration
	Out                 *os.File
	ErrOut              *os.File
	PodSelector    string
}

func RunNodeDrain(drainer *Drainer, nodeName string) error {
	pods, err := drainer.Client.CoreV1().Pods("").List(context.Background(), metav1.ListOptions{
		FieldSelector: "spec.nodeName=" + nodeName,
		LabelSelector: drainer.PodSelector
	})
```

**Review by andrewd-zededa — CHANGES_REQUESTED** [review]
Reviewed, I have a couple of change suggestions otherwise looks good.

File: pkg/pillar/cmd/zedkube/kubeinterface.go (hunk 4 above)
Comment (naiming-zededa) [reply pkg/pillar/cmd/zedkube/kubeinterface.go:129]: sure

File: pkg/pillar/cmd/zedkube/kubeinterface.go (hunk 5 above)
Comment (naiming-zededa) [reply pkg/pillar/cmd/zedkube/kubeinterface.go:100]: sounds good.

**Review by andrewd-zededa — APPROVED** [review]

**Pull request merged**